libopenage/renderer/opengl/renderer.cpp

```diff
@@ -72,14 +74,38 @@ resources::Texture2dData GlRenderer::display_into_data() {
 	return img.flip_y();
 }
 
-void GlRenderer::render(RenderPass const& pass) {
-	auto gl_target = dynamic_cast<const GlRenderTarget*>(pass.target);
+
+void GlRenderer::optimise(RenderPass& pass) {
+		if(!pass.get_is_optimised()) {
+			auto renderables = pass.get_renderables();
+			std::sort(renderables.begin(), renderables.end(), [](const Renderable& a, const Renderable& b) {
+				GLuint shader_a = dynamic_cast<GlUniformInput const*>(a.unif_in)->program->get_handle();
+				GLuint shader_b = dynamic_cast<GlUniformInput const*>(b.unif_in)->program->get_handle();
+
+				auto geom_a = dynamic_cast<GlGeometry const*>(a.geometry);
+				auto geom_b = dynamic_cast<GlGeometry const*>(b.geometry);
+				auto vao_a = (geom_a)? geom_a->get_vao_handle() : 0;
+				auto vao_b = (geom_b)? geom_b->get_vao_handle() : 0;
+				if (shader_a == shader_b)
+					return vao_a.value_or(0) < vao_b.value_or(0);
+				else
+					return shader_a < shader_b;
+			});
+
+			pass.set_renderables(renderables);
+			pass.set_is_optimised(true);
```

Comment:
Hm, I thought about it a bit more and it seems that reordering by VAOs is not actually a correct optimization - sorry about this! The reason is that uniform state of every shader is guaranteed to be preserved across a render pass. So for example, with the following renderables : "{set UNIF_A=1, draw VAO_1}, {set UNIF_A=2, draw VAO_2}, {draw VAO_1}", the last renderable can expect `UNIF_A` to be `2` and reordering would break that. So it seems the only valid thing is to do a `stable_sort` (`std::sort` is unstable) by shader handles, but not do anything with VAOs.